This file documents the `rds` files generated by Retroinspector.
They are all data.tables from the eponymous R package, but they can be coerced into base R data.frames or dplyr tibbles/data.frames.

# Insertions

## annotatedInsertionsMin3.<allPrefix>.rds

| Column name | Meaning | Data type |
| --- | --- | --- |
| seqnames | Chromosome | Factor with levels chr1-22,X,Y |
| start | INS position | int |
| end | INS position + 1 | int |
| width | end - start | int |
| strand | Strand in which the TE sequence was detected | Factor w/ 3 levels "+","-","*" |
| name | TE type, like _AluY_ | `chr` |
| score | Meaningless column | `int` |
| seqId | ID for each INS. They are unique for each run of the pipeline for other INS, but may (very likely) collide with deletions' IDs | `chr` |
| sw_score | RepeatMasker score | `int` |
| repeat.class | TE family, like `SINE` | `factor` |
| repeat.subclass | TE subfamily, like `Alu` | `factor` |
| repeat.start | Start of TE sequence within INS (from RepeatMasker) | `int` |
| repeat.end | End of TE sequence within INS (from RepeatMasker) | `int` |
| repeat.left | Bases left after TE sequence within INS (from RepeatMasker) | `int` |
| repeat.strains | Strand in which the TE sequence was detected (from RepeatMasker) | `chr` |
| repeat.divergence_percentage | Percentage of difference from of TE sequence (from RepeatMasker) | `num` |
| repeat.deletion_percentage | Percentage of bases deleted from of TE sequence (from RepeatMasker) | `num` |
| repeat.insertion_percentage | Percentage of bases inserted from of TE sequence (from RepeatMasker) | `num` |
| vcf_alt | Inserted sequence from the first sample in the cohort | `chr` |
| SUPP_VEC_min3 | String of 1s and 0s listing which samples carry and do not carry an INS, respectively, according to the lax criterion | `chr` |
| SUPP_min3 | Number of samples that carry an INS, according to the lax criterion | `int` |
| SUPP_VEC_trusty | String of 1s and 0s listing which samples carry and do not carry an INS, respectively, according to the strict criterion | `chr` |
| SUPP_trusty | Number of samples that carry an INS, according to the strict criterion | `int` |
| SVLEN | Length of the inserted sequence | `num` |
| genotype | For each INS, a numeric vector containing a number for the genotype of that INS on each sample. `0` means homozygous for the reference allele (no insertion), `1` means heterozygous, `2` means homozygous for the insertion. | `num` |
| SUPP_VEC_geno | String of 1s and 0s listing which samples carry and do not carry an INS, respectively, according to genotype | `chr` |
| SUPP_geno | Number of samples that carry an INS, according to genotype | `int` |
| maf | Allele fraction inside the sample (from 0 to 1) | `numeric` |
| support | For each insertion, the number of supporting reads in each sample | `numeric` of numeric vectors |
| SUPP_mask_min3 | For each insertion, TRUE if a sample carries it, FALSE if it doesn't, according to the lax criterion | `list` of logical vectors |
| SUPP_mask_trusty |  For each insertion, TRUE if a sample carries it, FALSE if it doesn't, according to the strict criterion | `list` of logical vectors |
| repeat.family | Classification of repeats, only applies to class II TEs (comes directly from RepeatMasker) | `chr` |
| annot.start | Start position of gene annotation | `int` |
| annot.end | End position of gene annotation | `int` |
| annot.gene_id | Gene ID of gene annotation | `chr` |
| annot.symbol | Gene name of gene annotation | `chr` |
| annot.class | Type of annotation, like `genes`. From annotatr | `chr` |
| annot.subclass | Type of annotation, like `exons`. From annotatr | `chr` |
| repeat.percentage | Percentage of INS covered by TE sequence | `num` |

## insertionsTable.<allPrefix>.rds

This contains the same columns than the previous table, except the ones related to genomic annotation. THe difference between both tables is that this one contains one row per INS, while `annotatedInsertionsMin3.<allPrefix>.rds` contains one row per annotation, resulting in multiple rows per INS.

## allIns.rds

This table contains the same columns than `annotatedInsertionsMin3.<allPrefix>.rds`. The difference is that this one contains INS with repetitive sequences that are not TEs, such as polyA.

# Deletions

## meDeletionsMin3.<allPrefix>.rds

| Column name | Meaning | Data type |
| --- | --- | --- |
| seqnames | Chromosome | `chr` |
| start | Starting position of DEL | `int` |
| end | Ending position of DEL | `num` |
| id |  ID for each DEL. They are unique for each run of the pipeline for other DEL, but may (very likely) collide with insertions' IDs | `chr` |
| SUPP_min3 | Number of samples that carry an INS, according to the lax criterion | `int` |
| SUPP_trusty | Number of samples that carry an INS, according to the strict criterion | `int` |
| genotype | For each DEL, a numeric vector containing a number for the genotype of that DEL on each sample. `0` means homozygous for the reference allele (no deletion), `1` means heterozygous, `2` means homozygous for the deletion. | `num` |
| SUPP_VEC_geno | String of 1s and 0s listing which samples carry and do not carry a DEL, respectively, according to genotype | `chr` |
| SUPP_geno | Number of patients that carry a DEL, according to genotyping, i.e., those that have a 0/1 or 1/1 genotype | `num` |
| maf | Allele frequency of the DEL (from 0 to 1) in the cohort on which the pipeline was run | `num` |
| svlen | Length of deletion | `num` |
| name | Type of TE sequence that was deleted, like _AluY_ | `chr` |
| repeat.class | Family of the deleted TE element, like _SINE_ | `chr` |
| repeat.coords | Coordinates of deleted TE sequences | `chr` |
| repeat.subclass | Subfamily of the deleted TE element, like _Alu_ | `chr` |
| repeat.family | Classification of repeats, only applies to class II TEs (comes directly from RepeatMasker) | `chr` |

# Enrichment

These files are produced by DOSE and clusterProfiler.
Please refer to the `enrichResult` documentation.

These files are `egoMF.rds`, `egoBP.rds`, and `egoCC.rds` for GO; and `do.rds`, `ncg.rds` for DO and NCG, respectiely.  
